core/src/test/java/com/netflix/iceberg/TestReplaceFiles.java

```diff
@@ -0,0 +1,244 @@
+        .apply();
+  }
+
+  @Test(expected = ValidationException.class)
+  public void testAddOnly() {
+    Assert.assertEquals("Table should start empty", 0, listMetadataFiles("avro").size());
+
+    table.newRewrite()
+            .replaceFiles(Sets.newSet(FILE_A), Collections.emptySet())
+            .apply();
+  }
+
+  @Test(expected = ValidationException.class)
+  public void testDeleteOnly() {
+    Assert.assertEquals("Table should start empty", 0, listMetadataFiles("avro").size());
+
+    table.newRewrite()
+            .replaceFiles(Collections.emptySet(), Sets.newSet(FILE_A))
+            .apply();
+  }
+
+  @Test
+  public void testAddAndDelete() {
+    Assert.assertEquals("Table should start empty", 0, listMetadataFiles("avro").size());
+
+    table.newAppend()
+            .appendFile(FILE_A)
+            .appendFile(FILE_B)
+            .commit();
+
+    TableMetadata base = readMetadata();
+    final long baseSnapshotId = base.currentSnapshot().snapshotId();
+    Assert.assertEquals("Should create 1 manifest for initial write",
+            1, base.currentSnapshot().manifests().size());
+    String initialManifest = base.currentSnapshot().manifests().get(0);
+
+    Snapshot pending = table.newRewrite()
+            .replaceFiles(Sets.newSet(FILE_A), Sets.newSet(FILE_C))
+            .apply();
+
+    Assert.assertEquals("Should contain 2 manifest",
+            2, pending.manifests().size());
+    Assert.assertNotEquals("Should not contain manifest from initial write",
+            initialManifest, pending.manifests().get(0));
+
+    long pendingId = pending.snapshotId();
+
+    validateManifest(pending.manifests().get(0),
+            ids(pendingId, baseSnapshotId),
+            concat(files(FILE_A, FILE_B)),
+            statuses(DELETED, EXISTING));
+
+    validateManifest(pending.manifests().get(1),
+            ids(pendingId),
+            concat(files(FILE_C)),
+            statuses(ADDED));
+  }
+
+  @Test
+  public void testFailure() {
+    table.newAppend()
+        .appendFile(FILE_A)
+        .commit();
+
+    table.ops().failCommits(5);
+
+    RewriteFiles rewrite = table.newRewrite()
+            .replaceFiles(Sets.newSet(FILE_A), Sets.newSet(FILE_B));
+    Snapshot pending = rewrite.apply();
+
+    Assert.assertEquals("Should produce 2 manifests", 2, pending.manifests().size());
+    String manifest1 = pending.manifests().get(0);
+    String manifest2 = pending.manifests().get(1);
+
+    validateManifest(manifest1, ids(pending.snapshotId()), concat(files(FILE_A)), statuses(DELETED));
+    validateManifest(manifest2, ids(pending.snapshotId()), concat(files(FILE_B)), statuses(ADDED));
+
+    AssertHelpers.assertThrows("Should retry 4 times and throw last failure",
+        CommitFailedException.class, "Injected failure", rewrite::commit);
+
+    Assert.assertFalse("Should clean up new manifest", new File(manifest1).exists());
+    Assert.assertFalse("Should clean up new manifest", new File(manifest2).exists());
```

Comment:
Apply is going to be called each time the commit is reattempted, so it would be good to also check that there aren't any unexpected manifests using `listMetadataFiles`.

Reply:
Added.